gmt-4.5.14/src/x2sys/x2sys_merge.1
.TH X2SYS_MERGE 1 "1 Nov 2015" "GMT 4.5.14" "Generic Mapping Tools"
.SH NAME
x2sys_merge \- Merge an updated COEs tables
.SH SYNOPSIS
\fBx2sys_merge\fP \fB\-A\fP\fImain_COElist.d\fP \fB\-M\fP\fInew_COElist.d\fP  
.SH DESCRIPTION
\fBx2sys_merge\fP will read two crossovers data base and output the contents
of the main one updated with the COEs in the second one. The second
file should only contain updated COEs relatively to the first one.
That is, it MUST NOT contain any new two tracks intersections (This point is
NOT checked in the code). This program is useful when, for any good reason like
file editing NAV correction or whatever, one had to recompute only the COEs between
the edited files and the rest of the database.
.TP
\fB\-A\fP
Specify the file \fImain_COElist.d\fP with the main crossover error data base.
.TP
\fB\-M\fP
Specify the file \fInew_COElist.d\fP with the newly computed crossover error data base.
.SH OPTIONS
No space between the option flag and the associated arguments.
.TP
.SH EXAMPLES
To update the main COE_data.txt with the new COEs estimations saved in the smaller COE_fresh.txt, try
.br
.sp
\fBx2sys_merge\fP \fB\-A\fP COE_data.txt \fB\-M\fP COE_fresh.txt > COE_updated.txt
.br
.sp
.SH "SEE ALSO"
.IR x2sys_binlist (1),
.IR x2sys_cross (1),
.IR x2sys_datalist (1),
.IR x2sys_get (1),
.IR x2sys_init (1),
.IR x2sys_list (1),
.IR x2sys_put (1),
.IR x2sys_report (1)


gmt-4.5.14/ChangeLog
-----------------------------------------------------------------------------------------
	$Id: ChangeLog 10331 2015-10-29 19:41:56Z pwessel $
-----------------------------------------------------------------------------------------

-----------------------------------------------------------------------------------------
2015-11-01	pwessel
	* GMT 4.5.14 finalized and released
-----------------------------------------------------------------------------------------
2015-10-29	pwessel
	* meca/psmeca.c:	Enforce rakes and strikes to be in -180/+180 range.
				Issue #799, for gmt4 only as not a problem in gmt5.

2015-07-28	pwessel
	* meca/psmeca.c:	Be able to read any value x and y coordinates

2015-06-24	pwessel
	* gmt_support.c:	Fix bug in jump near Dateline in gmt_inonout_sphpol_count,
				affecting inside/outside spherical cases, reported in message-2219

2015-03-18	pwessel
	* share/cpt/GMT_wysiwyg.cpt:	Fix typo in blue for high end [Issue #689].

2015-01-16	pwessel
	* gmt_plot.c:		Fix issue # 662 with graph arrows for negative scales.

2015-01-10	pwessel
	* gmt_plot.c:		Flipping the n/s sides array when lon and -y axes
				fixes issue #520

2015-01-01	pwessel
	* configure.ac:		Moved to 4.5.14 and back to svn release only.
-----------------------------------------------------------------------------------------
2015-01-01	pwessel
	* GMT 4.5.13 finalized and released [GSHHG 2.3.4 was released on Jan 1 as well]
-----------------------------------------------------------------------------------------
2014-12-28	pwessel
	* gmt_map.c:		Needed separate GMT_cartesian_overlap for non-geographic
				projections.  Fixes issue # 657.

2014-09-29	pwessel
	* gmt_plot.c:		Left a PSL variable undefined when grdimage plotted a grid
				with x = longitude and y = Cartesian.

2014-09-09	pwessel
	* grdfilter.c:		Fix y-shifts when -I sets smaller output.  Issue #616.
	* gmt_bcr.c:		Fix problem with points slightly outside causing an
				access violation; see issue #620.

2014-08-29	pwessel
	* sample1d.c:		Fix resampling for decreasing t values.  Issue #608.

2014-08-05	pwessel
	* gmt_map.c:		Could run into trouble clipping polygons for non-periodic
				maps.  Added more checks in GMT_wesn_clip.

2014-07-31	pwessel
	* gmt_map.c:		Rect clipping with -R<w/s/e/n>r could introduce stray lines
				when dataset has features about 180 degrees away from
				the user's area. See issue # 594.

2014-07-30	pwessel
	* configure.ac:		Set default GSHHG version to 2.3.2 (released August 1, 2014)
	* psxy[z].c:		Fixed wrap-around issue in fault lines.

2014-07-27	pwessel
	* gmtselect.c:		The -Z option needed to pass records with NaNs.

2014-07-19	pwessel
	* x2sys/x2sys_solve.c:	Fix problems that arise when a track has less crossovers
				than model parameters.
	* gmtmath.c:		Fix issue #123 with files and -C.

2014-07-01	pwessel
	* configure.ac:		Set default GSHHG version to 2.3.1 (released July 1, 2014)
	
2014-06-28	pwessel
	* mgd77/mgd77sniffer.c:	Needed to check that grid values used in decimate function
	 			also were within reasonable range.

2014-05-26	pwessel
	* gmt_map.c:		The map region clip path for -JE/e was hardwired to a full
				circle and did not care about -R setting.
				When -R...r and -Je/E was given the search for enclosing
				geographic boundaries failed when either the S or N pole
				was the projection's antipole.
2014-05-25	pwessel
	* grdmath.c:		Fix issue #566 bug in grdmath's KM2DEG operator.

2014-03-19	pwessel
	* gmtselect.c:		The -Fpolygon option might miss points because the io-machinery
				would sometimes set the min/max longitudes found incorrectly.

2014-03-14	pwessel
	* gmt_shore.c:		Give more error feedback if nc_open fails due to old library.

2014-03-01	pwessel
	* configure.ac:		Moved to 4.5.13 and back to svn release only.

-----------------------------------------------------------------------------------------
2014-03-01	pwessel
	* GMT 4.5.12 finalized and released [GSHHG 2.3.0 was released on Feb 1]
-----------------------------------------------------------------------------------------
2014-02-17	fwobbe
	* gmt_init.c etc.:	File locking -- if enabled via -DFLOCK -- is
				non-compulsory now.  This means if fcntl fails for some reason,
				GMT continues to access the file.  Hence, it is recommended to
				configure GMT with --enable-flock even if fcntl might be buggy
				on some ancient NFS shares.

2014-02-17	pwessel
	* gmt_map.c:		Size for memory allocation cast the product of nx * ny to
				size_t but overflow had already happen as calculation was
				done using ints, then cast to size_t. Must cast nx and ny
				separately.
2014-02-10	jluis
	* misc/gmt2kml.c:	Make sure the default style IDs are unique for each process.

2014-02-07	jluis
	* ps2raster.c:		Make sure that images in KML files do not carry a directory name

2014-02-03	pwessel
	* pslib.c:		Auto-widen paragraph width to at least fit widest word.

2014-02-01	pwessel
	* configure.ac:		Since GSHHG 2.3.0 was released today, make that the default

2014-01-26	pwessel
	* gmt_grdio.c:		Building the grid command string could exceed its length if
				the argument became exactly the max length of the array; then
				a final extra character \0 would be added and exceed array length.
				Thanks to Joachim Saul for pointing it out.

2014-01-22	pwessel
	* pscoast.c:		The processing of -I -N levels and pens could get out of sync
				so that the wrong pen was used for the specified feature.

2014-01-16	remko
	* pslib.c:		Take superscripts and subscripts into account when determining
				dimensions of text box.

2014-01-10	pwessel
	* grdmath.c:		Had to let SDIST return km instead of degrees since it cannot
				do degrees for geodesics.  By doing km it is now aligned with
				what PDIST and LDIST returns for geographic data, and matches
				what GMT5 does.  Some scripts had to be modified to handle the
				new behavior.  If you use SDIST you may wish to check your usage.
				Also fixed bug in SDIST in that spherical selected geodesics

gmt-4.5.14/src/x2sys/x2sys_merge.c
/*-----------------------------------------------------------------
 *	$Id: x2sys_merge.c 10289 2014-12-28 21:17:06Z pwessel $
 *
 *      Copyright (c) 1999-2015 by J. Luis
 *      See LICENSE.TXT file for copying and redistribution conditions.
 *
 *      This program is free software; you can redistribute it and/or modify
 *      it under the terms of the GNU General Public License as published by
 *      the Free Software Foundation; version 2 or any later version.
 *
 *      This program is distributed in the hope that it will be useful,
 *      but WITHOUT ANY WARRANTY; without even the implied warranty of
 *      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *      GNU General Public License for more details.
 *
 *      Contact info: gmt.soest.hawaii.edu
 *--------------------------------------------------------------------*/
/* x2sys_merge will read two crossovers data base and output the contents
 * of the main one updated with the COEs in the second one. The second
 * file should only contain updated COEs relatively to the first one.
 * That is, it MUST NOT contain any new two tracks intersections. 
 *
 * Author:	Joaquim Luis
 * Date:	09-Juin-2009
 *
 */

#include "gmt.h"

int main (int argc, char **argv) {

	int  i, j, k, n_alloc, n_base, n_merge, *map_base_start = NULL, *map_base_end = NULL, *map_merge_start = NULL, *map_merge_end = NULL;
	int  merge_start;
	char line[BUFSIZ], *dbase1 = CNULL, *dbase2 = CNULL, **pairs_base = NULL, **pairs_merge = NULL, *c_not_used = NULL;
	FILE *fp_base = NULL, *fp_merge = NULL;

	for (i = 1; i < argc; i++) {
		if (argv[i][0] == '-') {
			switch (argv[i][1]) {
				case 'A':
					dbase1 = &argv[i][2];
					break;
				case 'M':
					dbase2 = &argv[i][2];
					break;
				default:
					GMT_give_synopsis_and_exit = 1;
					break;
			}
		}
	}

	if (argc == 1 || GMT_give_synopsis_and_exit) {
		fprintf (stderr, "x2sys_merge - Merge an updated COEs table (smaller) into a main one (bigger)\n\n");
		fprintf (stderr, "usage: x2sys_merge -A<main_COEdbase> -M<new_COEdbase>\n");

		if (GMT_give_synopsis_and_exit) exit (EXIT_FAILURE);

		fprintf (stderr, "\t-A<main_COEdbase> File with the main crossover error data base.\n");
		fprintf (stderr, "\t-M<new_COEdbase> File with the new crossover error data base.\n");
		fprintf (stderr, "\t  The new COEs will replace the old ones present in <main_COEdbase>.\n\n");
		fprintf (stderr, "\t  Result is printed to stdout.\n");
		exit (EXIT_FAILURE);
	}

	if (!dbase1) {
		fprintf (stderr, "%s: ERROR: Missing Base COEs database file. -A is mandatory\n", GMT_program);
		exit (EXIT_FAILURE);
	}

	if (!dbase2) {
		fprintf (stderr, "%s: ERROR: Missing Updating COEs database file. -M is mandatory\n", GMT_program);
		exit (EXIT_FAILURE);
	}

	if ((fp_base = fopen (dbase1, "r")) == NULL) {
		fprintf (stderr, "%s: ERROR: Unable to open crossover file %s\n", GMT_program, dbase1);
		exit (EXIT_FAILURE);
	}

	if ((fp_merge = fopen (dbase2, "r")) == NULL) {
		fprintf (stderr, "%s: ERROR: Unable to open crossover file %s\n", GMT_program, dbase2);
		exit (EXIT_FAILURE);
	}

	n_alloc = GMT_CHUNK;
	map_base_start = (int *) GMT_memory (VNULL, (size_t)n_alloc, sizeof (int), GMT_program);
	map_base_end = (int *) GMT_memory (VNULL, (size_t)n_alloc, sizeof (int), GMT_program);
	pairs_base = (char **) GMT_memory (VNULL, (size_t)n_alloc, sizeof (char *), GMT_program);

	map_merge_start = (int *) GMT_memory (VNULL, (size_t)n_alloc, sizeof (int), GMT_program);
	map_merge_end = (int *) GMT_memory (VNULL, (size_t)n_alloc, sizeof (int), GMT_program);
	pairs_merge = (char **) GMT_memory (VNULL, (size_t)n_alloc, sizeof (char *), GMT_program);

	/* Read in the main COEs dbase and store the pair track names */
	n_base = 0;		k = 1;
	while (fgets (line, BUFSIZ, fp_base)) {	/*  */
		if (line[0] == '>') {
			map_base_start[n_base] = k;
			if (n_base) map_base_end[n_base-1] = k - 1;
			pairs_base[n_base] = (char *) GMT_memory (VNULL, (size_t)24, sizeof (char), GMT_program);
			strncpy(pairs_base[n_base], &line[2], 19);
			n_base++;
			if (n_base == n_alloc) {
				n_alloc <<= 1;
				map_base_start = (int *) GMT_memory ((void *)map_base_start, (size_t)n_alloc, sizeof (int), GMT_program);
				map_base_end = (int *) GMT_memory ((void *)map_base_end, (size_t)n_alloc, sizeof (int), GMT_program);
				pairs_base = (char **) GMT_memory ((void *)pairs_base, (size_t)n_alloc, sizeof (char *), GMT_program);
			}
		}

		k++;
	}
	map_base_end[n_base - 1] = k - 1;	/* This one was not yet assigned */
	rewind(fp_base);

	/* Read in the updated COEs dbase and store the pair track names */
	n_alloc = GMT_CHUNK;
	n_merge = 0;		k = 1;
	while (fgets (line, BUFSIZ, fp_merge)) {	/*  */
		if (line[0] == '>') {
			map_merge_start[n_merge] = k;
			if (n_merge) map_merge_end[n_merge-1] = k - 1;
			pairs_merge[n_merge] = (char *) GMT_memory (VNULL, (size_t)24, sizeof (char), GMT_program);
			strncpy(pairs_merge[n_merge], &line[2], 19);
			n_merge++;
			if (n_merge == n_alloc) {
				n_alloc <<= 1;
				map_merge_start = (int *) GMT_memory ((void *)map_merge_start, (size_t)n_alloc, sizeof (int), GMT_program);
				map_merge_end = (int *) GMT_memory ((void *)map_merge_end, (size_t)n_alloc, sizeof (int), GMT_program);
				pairs_merge = (char **) GMT_memory ((void *)pairs_merge, (size_t)n_alloc, sizeof (char *), GMT_program);
			}
		}

		k++;
	}
	map_merge_end[n_merge - 1] = k - 1;	/* This one was not yet assigned */
	rewind(fp_merge);


	/* Jump comment lines in both files and osition the file poiter into the first data line */
	k = 0;
	while (fgets (line, BUFSIZ, fp_merge) && line[0] == '#') k++;	/* Jump the comment lines in the to-merge file */
	rewind(fp_merge);
	for (i = 0; i < k; i++) c_not_used = fgets (line, BUFSIZ, fp_merge);

	k = 0;
	while (fgets (line, BUFSIZ, fp_base) && line[0] == '#') {
		fprintf (stdout, "%s", line);
		k++;
	}
	rewind(fp_base);
	for (i = 0; i < k; i++) c_not_used = fgets (line, BUFSIZ, fp_base);

	/* Do the merging. COEs present in file dbase2 replace their pairs in dbase1 */
	merge_start = 0;
	for (i = 0; i < n_base; i++) {
		for (j = merge_start; j < n_merge; j++) {
			if (!strcmp(pairs_base[i], pairs_merge[j])) {		 /* Update these COEs */
				for (k = map_merge_start[j]; k <= map_merge_end[j]; k++) {
					c_not_used = fgets (line, BUFSIZ, fp_merge);
					fprintf (stdout, "%s", line);
				}
				for (k = map_base_start[i]; k <= map_base_end[i]; k++)	/* Advance also in the base file */
					c_not_used = fgets (line, BUFSIZ, fp_base);

				merge_start = j + 1;
				break;		/* Since we found this to update, no need to continue seeking for another repetition */
			}
			else if (j == (n_merge - 1)) {	/* Not equal. So do not to update, just recopy */
				for (k = map_base_start[i]; k <= map_base_end[i]; k++) {
					c_not_used = fgets (line, BUFSIZ, fp_base);
					fprintf (stdout, "%s", line);
				}
			}
		}
		if (merge_start == n_merge) {	/* Copy the rest of dbase1 file and stop */
			while (fgets (line, BUFSIZ, fp_base))
				fprintf (stdout, "%s", line);
			break;			/* Not very elegant way of stopping, but we are done */
		}
	}

	fclose(fp_base);
	fclose(fp_merge);

	for (i = 0; i < n_base; i++) GMT_free ((void *)pairs_base[i]);
	for (i = 0; i < n_merge; i++) GMT_free ((void *)pairs_merge[i]);
	GMT_free((void *) map_base_start);
	GMT_free((void *) map_base_end);
	GMT_free((void *) map_merge_start);
	GMT_free((void *) map_merge_end);

	exit (EXIT_SUCCESS);
}
